client/db/bolt/db.go

```diff
@@ -163,14 +164,55 @@ func NewDB(dbPath string, logger dex.Logger) (dexdb.DB, error) {
 	return bdb, bdb.upgradeDB()
 }
 
+func (db *BoltDB) fileSize(path string) int64 {
+	stat, err := os.Stat(path)
+	if err != nil {
+		db.log.Errorf("Failed to stat %v: %v", path, err)
+		return 0
+	}
+	return stat.Size()
+}
+
 // Run waits for context cancellation and closes the database.
 func (db *BoltDB) Run(ctx context.Context) {
-	<-ctx.Done()
+	<-ctx.Done() // wait for shutdown to backup and compact
+
+	// Create a backup in the backups folder.
+	db.log.Infof("Backing up database...")
```

Comment:
I'm doing it this way with a vanilla backup first in case the compacted file has an issue.  But this could be changed to just to a BackupTo with compaction to the "backups" folder, and then copy the resulting file over the source file after closing the source db.